pkg/agent/controller/bgp/controller.go

```diff
@@ -0,0 +1,954 @@
+	ipv6Suffix = "/128"
+)
+
+const key = "dummyKey"
+
+type bgpPolicyState struct {
+	// The name of BGPPolicy takes effect.
+	bgpPolicy string
+	// The local BGP server instance.
+	bgpServer bgp.Interface
+	// The port on which the local BGP server listens.
+	listenPort int32
+	// The AS number used by the local BGP server.
+	localASN int32
+	// The router ID used by the local BGP server.
+	routerID string
+	// routes stores all BGP routers advertised to BGP peers.
+	routes sets.Set[bgp.Route]
+	// peerConfigs is a map that stores configurations of BGP peers. The map keys are the concatenated strings of BGP
+	// peer IP address and ASN (e.g., "192.168.77.100-65000", "2001::1-65000").
+	peerConfigs map[string]bgp.PeerConfig
+}
+
+type Controller struct {
+	ctx context.Context
+
+	nodeInformer     cache.SharedIndexInformer
+	nodeLister       corelisters.NodeLister
+	nodeListerSynced cache.InformerSynced
+
+	serviceInformer     cache.SharedIndexInformer
+	serviceLister       corelisters.ServiceLister
+	serviceListerSynced cache.InformerSynced
+
+	egressInformer     cache.SharedIndexInformer
+	egressLister       crdlistersv1b1.EgressLister
+	egressListerSynced cache.InformerSynced
+
+	bgpPolicyInformer     cache.SharedIndexInformer
+	bgpPolicyLister       crdlistersv1a1.BGPPolicyLister
+	bgpPolicyListerSynced cache.InformerSynced
+
+	endpointSliceInformer     cache.SharedIndexInformer
+	endpointSliceLister       discoverylisters.EndpointSliceLister
+	endpointSliceListerSynced cache.InformerSynced
+
+	bgpPolicyState *bgpPolicyState
+
+	k8sClient             kubernetes.Interface
+	bgpPeerPasswords      map[string]string
+	bgpPeerPasswordsMutex sync.RWMutex
+
+	nodeName     string
+	enabledIPv4  bool
+	enabledIPv6  bool
+	podIPv4CIDR  string
+	podIPv6CIDR  string
+	nodeIPv4Addr string
+
+	egressEnabled bool
+
+	newBGPServerFn func(globalConfig *bgp.GlobalConfig) bgp.Interface
+
+	queue workqueue.RateLimitingInterface
+}
+
+func NewBGPPolicyController(ctx context.Context,
+	nodeInformer coreinformers.NodeInformer,
+	serviceInformer coreinformers.ServiceInformer,
+	egressInformer crdinformersv1b1.EgressInformer,
+	bgpPolicyInformer crdinformersv1a1.BGPPolicyInformer,
+	endpointSliceInformer discoveryinformers.EndpointSliceInformer,
+	k8sClient kubernetes.Interface,
+	nodeConfig *config.NodeConfig,
+	networkConfig *config.NetworkConfig) (*Controller, error) {
+	c := &Controller{
+		ctx:                       ctx,
+		nodeInformer:              nodeInformer.Informer(),
+		nodeLister:                nodeInformer.Lister(),
+		nodeListerSynced:          nodeInformer.Informer().HasSynced,
+		serviceInformer:           serviceInformer.Informer(),
+		serviceLister:             serviceInformer.Lister(),
+		serviceListerSynced:       serviceInformer.Informer().HasSynced,
+		egressInformer:            egressInformer.Informer(),
+		egressLister:              egressInformer.Lister(),
+		egressListerSynced:        egressInformer.Informer().HasSynced,
```

Comment:
This should be executed only when egress is anebld, otherwise it would start an informer

Reply:
Done